File: r/02.R
library(tidyverse)

# Part One ------------------------------------------------------------------

input <- read_delim("raw/02.txt", delim = " ", col_names = FALSE)

input_clean <- input %>% 
  separate(X1, c("min_n", "max_n"), sep = "-") %>% 
  rename(letter = X2, pwd = X3) %>% 
  mutate(letter = str_remove(letter, ":")) 

all_passwords <- input_clean %>% 
  mutate(letter_n = str_count(pwd, letter)) %>% 
  rowwise() %>% 
  mutate(is_valid = between(letter_n, min_n, max_n))

sum(all_passwords$is_valid)


# Part Two -----------------------------------------------------------------

all_passwords_rule2 <- input_clean %>% 
  mutate(pos1 = substr(pwd, min_n, min_n),
         pos2 = substr(pwd, max_n, max_n)) %>% 
  mutate(match_n = (letter == pos1) + (letter == pos2))

sum(all_passwords_rule2$match_n == 1)

File: py/02.py
import re

file = open("../raw/02.txt")

input = file.read()

input_lst = input.split("\n")

all_inputs_split = [re.split('[ :-]', i) for i in input_lst]

# Part One

pwd_check = [int(i[0]) <=  i[4].count(i[2]) <= int(i[1]) for i in all_inputs_split if len(i) == 5]

sum(pwd_check)

# Part Two

test_cases = [
    ["1", "3", "a", "", "abcde"],
    ["1", "3", "b", "", "cdefg"],
    ["2", "9", "c", "", "ccccccccc"]
]

def apply_rule(x):
    pos1 = int(x[0])
    pos2 = int(x[1])
    ltr = x[2]
    pwd = x[4]

    check_one = pwd[pos1 - 1] == ltr
    if pos2 <= len(pwd):
        check_two = pwd[pos2 - 1] == ltr
    else:
        check_two = False

    #print(check_one, check_two)
    return check_one + check_two == 1

[apply_rule(i) for i in test_cases if len(i) == 5]

sum([apply_rule(i) for i in all_inputs_split if len(i) == 5])
